vnext/ReactUWP/Modules/Animated/DecayAnimationDriver.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "DecayAnimationDriver.h"
+#include <jsi/jsi.h>
+#include <math.h>
+
+namespace react { namespace uwp {
+  DecayAnimationDriver::DecayAnimationDriver(int64_t id, const std::shared_ptr<ValueAnimatedNode>& animatedValue, const Callback& endCallback, const folly::dynamic& config) : AnimationDriver(id, animatedValue, endCallback)
+  {
+    m_deceleration = config.find("deceleration").dereference().second.asDouble();
+    m_velocity = config.find("velocity").dereference().second.asDouble();
+    m_iterations = [iterations = config.find("iterations"), end = config.items().end()]() {
+      if (iterations != end)
+      {
+        return iterations.dereference().second.getInt();
+      }
+      return static_cast<int64_t>(1);
+    }();
+
+    const auto [animation, scopedBatch] = [deceleration = m_deceleration, velocity = m_velocity, iterations = m_iterations, fromValue = m_animatedValue->RawValue()]()
+    {
+      const auto [scopedBatch, animation] = []()
+      {
+        const auto compositor = winrt::Window::Current().Compositor();
+        return std::make_tuple(compositor.CreateScopedBatch(winrt::CompositionBatchTypes::AllAnimations), compositor.CreateScalarKeyFrameAnimation());
+      }();
+
+      std::chrono::milliseconds duration(static_cast<int>(velocity / -deceleration * 1000));
+      animation.Duration(duration);
+
+      const auto compositor = winrt::Window::Current().Compositor();
+      animation.SetScalarParameter(L"velocity", static_cast<float>(velocity));
+      animation.SetScalarParameter(L"deceleration", static_cast<float>(deceleration));
+      animation.SetScalarParameter(L"duration", static_cast<float>(velocity / -deceleration) * 1000);
+      animation.InsertExpressionKeyFrame(1.0f, L"(duration * velocity) + (0.5 * deceleration * duration * duration)", compositor.CreateCubicBezierEasingFunction({ 0,1 }, {0, 1}));
+      animation.IterationCount(static_cast<int32_t>(iterations));
+      animation.IterationBehavior(winrt::AnimationIterationBehavior::Count);
+
+      return std::make_tuple(animation, scopedBatch);
+    }();
+
+    if (m_animatedValue)
+    {
+      m_animatedValue->PropertySet().StartAnimation(L"offset", animation);
+      m_animatedValue->AddActiveAnimation(m_id);
+    }
+    scopedBatch.End();
+
+    scopedBatch.Completed([endCallback = m_endCallback, animatedValue = m_animatedValue, id = m_id](auto sender, auto)
```

Comment:
>[endCallback = m_endCallback, animatedValue = m_animatedValue, id = m_id] [](start = 26, length = 73)

Is there a lifetime issue with these captured variables?  eg is it possible to destroy the animation driver before the scoped batch completes?

Reply:
Yes, in fact it happens quite often, which is why I've had to capture these values via a copy constructor rather then referencing the originals.  Do you object to this?

---
In reply to: [296939653](https://github.com/microsoft/react-native-windows/pull/2556#discussion_r296939653) [](ancestors = 296939653)

Reply:
I think I would prefer to unregister the scoped batch completion handler when the AnimationDriver is being deleted (we do this in WUX facades code).  Having said that, as long as there isn't a runtime issue I'd be fine with tackling this later.